drivers/usb/udc/udc_it82xx2.c

```diff
@@ -0,0 +1,1434 @@
+		if (udc_ctrl_stage_is_status_out(dev) && len != 0) {
+			LOG_DBG("Handle early setup token");
+			/* Notify upper layer */
+			udc_ctrl_submit_status(dev, buf);
+			/* Update to next stage of control transfer */
+			udc_ctrl_update_stage(dev, buf);
+			return 0;
+		}
+	}
+
+	ep_cfg = udc_get_ep_cfg(dev, ep);
+	if (len > ep_cfg->mps) {
+		LOG_ERR("Failed to handle this packet due to the packet size");
+		return -ENOBUFS;
+	}
+
+	it82xx2_xfer_out_data(dev, ep, buf);
+
+	if (ep == USB_CONTROL_EP_OUT) {
+		if (udc_ctrl_stage_is_status_out(dev)) {
+			/* Status stage finished, notify upper layer */
+			udc_ctrl_submit_status(dev, buf);
+		}
+
+		/* Update to next stage of control transfer */
+		udc_ctrl_update_stage(dev, buf);
+
+		if (udc_ctrl_stage_is_status_in(dev)) {
+			it82xx2_usb_set_ep_ctrl(dev, ep, EP_DATA_SEQ_1, true);
+			err = udc_ctrl_submit_s_out_status(dev, buf);
+		}
+	} else {
+		err = udc_submit_ep_event(dev, buf, 0);
+	}
+
+	return err;
+}
+
+static void xfer_work_handler(const struct device *dev)
+{
+	while (true) {
+		struct it82xx2_ep_event evt;
+		int err = 0;
+
+		k_msgq_get(&evt_msgq, &evt, K_FOREVER);
+
+		switch (evt.event) {
+		case IT82xx2_EVT_SETUP_TOKEN:
+			err = work_handler_setup(evt.dev, evt.ep);
+			break;
+		case IT82xx2_EVT_IN_TOKEN:
+			err = work_handler_in(evt.dev, evt.ep);
+			break;
+		case IT82xx2_EVT_OUT_TOKEN:
+			err = work_handler_out(evt.dev, evt.ep);
+			break;
+		case IT82xx2_EVT_XFER:
+			break;
+		default:
+			LOG_ERR("Unknown event type 0x%x", evt.event);
+			err = -EINVAL;
+			break;
+		}
+
+		if (err) {
+			udc_submit_event(evt.dev, UDC_EVT_ERROR, err);
+		}
+
+		if (evt.ep != USB_CONTROL_EP_OUT && !udc_ep_is_busy(evt.dev, evt.ep)) {
+			if (work_handler_xfer_next(evt.dev, evt.ep) == 0) {
+				udc_ep_set_busy(evt.dev, evt.ep, true);
+			}
+		}
+	}
+}
+
+static void it82xx2_usb_xfer_done(const struct device *dev)
+{
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct it82xx2_usb_ep_regs *ep_regs = usb_regs->usb_ep_regs;
+	struct it82xx2_usb_ep_fifo_regs *ff_regs = usb_regs->fifo_regs;
+	struct epn_ext_ctrl_regs *epn_ext_ctrl =
+		usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
+
+	for (uint8_t fifo_idx = 0; fifo_idx < 4; fifo_idx++) {
+		uint8_t ep_ctrl = ep_regs[fifo_idx].ep_ctrl.value;
+
+		/* check ready bit ,will be 0 when trans done */
+		if ((ep_ctrl & ENDPOINT_EN) && !(ep_ctrl & ENDPOINT_RDY)) {
+			uint8_t ep, ep_idx;
+
+			switch (ep_regs[fifo_idx].ep_transtype_sts & DC_ALL_TRANS) {
+			case DC_SETUP_TRANS:
+				/* SETUP transaction done */
+				if (fifo_idx == 0) {
+					ep = 0x0;
+					ep_idx = USB_EP_GET_IDX(ep);
+					if (it82xx2_usb_set_ep_ctrl(dev, ep, EP_STALL_CHECK,
+								    true)) {
+						it82xx2_usb_set_ep_ctrl(dev, ep, EP_STALL_SEND,
+									false);
+						ff_regs[ep_idx].ep_rx_fifo_ctrl = FIFO_FORCE_EMPTY;
+						LOG_ERR("Cleared stall bit and rx fifo");
+						break;
+					}
```

Comment:
If you are indenting for the fourth time, you should look for a way to refactor your code. You can move this block to a inline function (same for the block below), `if (fifo_idx == 0)` is unlikely necessary.

Reply:
the code is updated. please help review it again.